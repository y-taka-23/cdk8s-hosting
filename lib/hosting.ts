import { Construct } from 'constructs';
import { ConfigMap, Container, Deployment, Volume } from 'cdk8s-plus';
import { Ingress } from '../imports/k8s';
import * as path from 'path';

export interface HostingOptions {
    host?: string
    public?: string
}

export class Hosting extends Construct {
  constructor(scope: Construct, name: string, opts: HostingOptions = {}) {
    super(scope, name);

    let publicDir = opts.public || 'public';
    if (!path.isAbsolute(publicDir)) {
        publicDir = path.join(path.dirname(process.argv[1]), publicDir);
    }

    const configMap = new ConfigMap(this, 'cm');
    configMap.addDirectory(publicDir);

    const port = 80;

    const container = new Container({
      image: 'nginx:1.19.2',
      port
    });
    container.mount('/usr/share/nginx/html', Volume.fromConfigMap(configMap));

    const deploy = new Deployment(this, 'deploy', {
      spec: {
        podSpecTemplate: {
          containers: [ container ]
        }
      }
    });

    const service = deploy.expose({ port });

    if (!opts.host) {
      new Ingress(this, 'ing', {
        spec: { backend: { serviceName: service.name, servicePort: port } }
      });
    } else {
      new Ingress(this, 'ing', {
        spec: {
          rules: [{
            host: opts.host,
            http: {
              paths: [{
                path: '/',
                backend: { serviceName: service.name, servicePort: port }
              }]
            }
          }]
        }
      });
    }
  }
}
