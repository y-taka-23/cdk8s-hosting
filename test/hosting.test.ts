import { Testing, Chart } from 'cdk8s';
import { Hosting } from '../lib/hosting';

test('default options', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'test');

  new Hosting(chart, 'mypage');

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});
