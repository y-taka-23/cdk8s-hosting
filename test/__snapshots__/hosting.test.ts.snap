// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with host option 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "index.html": "<!DOCTYPE html>
<html>
  <head>
    <title>cdk8s Demo</title>
  </head>
  <body>
    <h1>Hello, CDK for Kubernetes!</h1>
  </body>
</html>
",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "test-mypage-cm3eba6298-configmap-948e51d1",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-mypage-deploy-pod-b0bcacb0",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.deployment": "testmypagedeployD20B4760",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.deployment": "testmypagedeployD20B4760",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [],
              "image": "nginx:1.19.2",
              "name": "main",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
              "volumeMounts": Array [
                Object {
                  "mountPath": "/usr/share/nginx/html",
                  "name": "configmap-test-mypage-cm3eba6298-configmap-948e51d1",
                },
              ],
            },
          ],
          "volumes": Array [
            Object {
              "configMap": Object {
                "name": "test-mypage-cm3eba6298-configmap-948e51d1",
              },
              "name": "configmap-test-mypage-cm3eba6298-configmap-948e51d1",
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "test-mypage-deploy-service-pod-b0fcf028",
    },
    "spec": Object {
      "externalIPs": Array [],
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "cdk8s.deployment": "testmypagedeployD20B4760",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "name": "test-mypage-ing-5309310c",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "www.example.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "test-mypage-deploy-service-pod-b0fcf028",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
    },
  },
]
`;

exports[`with public option 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "index.html": "<!DOCTYPE html>
<html>
  <head>
    <title>cdk8s Demo</title>
  </head>
  <body>
    <h1>Hello, CDK for Kubernetes!</h1>
  </body>
</html>
",
    },
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "test-mypage-cm3eba6298-configmap-948e51d1",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-mypage-deploy-pod-b0bcacb0",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.deployment": "testmypagedeployD20B4760",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.deployment": "testmypagedeployD20B4760",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [],
              "image": "nginx:1.19.2",
              "name": "main",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
              "volumeMounts": Array [
                Object {
                  "mountPath": "/usr/share/nginx/html",
                  "name": "configmap-test-mypage-cm3eba6298-configmap-948e51d1",
                },
              ],
            },
          ],
          "volumes": Array [
            Object {
              "configMap": Object {
                "name": "test-mypage-cm3eba6298-configmap-948e51d1",
              },
              "name": "configmap-test-mypage-cm3eba6298-configmap-948e51d1",
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "test-mypage-deploy-service-pod-b0fcf028",
    },
    "spec": Object {
      "externalIPs": Array [],
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "cdk8s.deployment": "testmypagedeployD20B4760",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "name": "test-mypage-ing-5309310c",
    },
    "spec": Object {
      "backend": Object {
        "serviceName": "test-mypage-deploy-service-pod-b0fcf028",
        "servicePort": 80,
      },
    },
  },
]
`;
